### PR TITLE
bpo-32309: Implement asyncio.to_thread()

### PR DESCRIPTION
**Pull request by aeros**
Implements `asyncio.to_thread`, a coroutine for asynchronously running IO-bound functions in a separate thread without blocking the event loop. See the discussion starting from [here](https://github.com/python/cpython/pull/18410#issuecomment-628930973) in GH-18410 for context. 

https://bugs.python.org/issue32309

Automerge-Triggered-By: @aeros

### CODE REVIEW

File: Doc/library/asyncio-task.rst
Comment (aeros) [inline Doc/library/asyncio-task.rst:658]: Added as a note based on [feedback](https://github.com/python/cpython/pull/18410#issuecomment-624616980) from @tmewett.

File: Doc/library/asyncio-task.rst
Comment (cjerdonek) [inline Doc/library/asyncio-task.rst:610]: I would also document that the `*args` and `**kwargs` are passed to `func` (as opposed to being reserved for modifying `to_thread()`'s behavior), and also in the docstring below.

Comment (aeros) [reply Doc/library/asyncio-task.rst:610]: That's a good idea, thanks.

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
This looks great to me. Please don't merge before @asvetlov and @cjerdonek approve.

File: Doc/library/asyncio-task.rst
Comment (1st1) [inline Doc/library/asyncio-task.rst:608]: You also need to add this function to the "high level APIs index"

**Comment by tmewett** [comment]
Is the name finalised? I saw the discussion about `run_in_thread` being a bit ambiguous but personally I don't think `to_thread` is very clear either.

I would suggest `create_thread`. Then, this would match up almost exactly with the semantics of `asyncio.create_task`

create_task: Wrap the coroutine into a Task and schedule its execution. Return the Task object.

create_thread: Wrap the function in a thread and start its execution. Return a future object.

If you see what I mean? The functions are very similar but one is for coros concurrently in the event loop and the other is for functions concurrently in threads.

**Review by cjerdonek — APPROVED** [review]

**Comment by cjerdonek** [comment]
Thanks. @tmewett I think the proposed name is good enough. Re: `create_thread()`, I think it has the issue that when backed by a pool, it's not necessarily creating a new thread. With `to_thread()`, you can think of it as "going to" a thread in the pool.

File: Doc/library/asyncio-task.rst
Comment (asvetlov) [inline Doc/library/asyncio-task.rst]: The documentation is often used as a source for copy-paste without careful reading.
I think adding a wrong code snippet without a very special markup (like red never-do-it text label) can be confusing.

I suggest keeping the only correct code example. Adding `.. warning::` section with a caution about blocking code inside a coroutine might be a good idea also.

**Review by asvetlov — APPROVED** [review]
I have a little doubt for the documentation text; the code itself and test are fine.

**Comment by 1st1** [comment]
@aeros Looks like everybody approves this so you can go ahead and merge!

Next todo:

* Implement `asyncio.from_thread`
* Implement auto-scaling of the default thread pool.

**Comment by aeros** [comment]
> @aeros Looks like everybody approves this so you can go ahead and merge!

Awesome, thanks for the reviews @1st1, @cjerdonek, and @asvetlov!

I'll discuss the documentation a bit more w/ Andrew regarding the examples, and then proceed w/ merging it.

File: Doc/library/asyncio-task.rst
Comment (aeros) [reply Doc/library/asyncio-task.rst:608]: Will do, thanks for the reminder.

File: Doc/library/asyncio-task.rst
Comment (aeros) [reply Doc/library/asyncio-task.rst]: > The documentation is often used as a source for copy-paste without careful reading.

Great point, I had not honestly not considered that users could potentially copy-paste that example without reading the context.

However, I do think that there's some significant value in the time comparisons. I largely credit the `create_task()` examples w/ time comparisons at the start of the page for greatly improving my own understanding of how tasks work. Especially with something like async, it's not always easy to grasp without some visualization.

I think what I'll do is remove the first example, and leave an after remark about the current second example saying something like:

"Directly calling `blocking_io()` would block the event loop for its duration, resulting in an additional 1 second of run time. Instead, by using `asyncio.to_thread()`, we can run it in a separate thread without blocking the event loop."

(I'll also of course revise the paragraph before the example.)

Thanks for the feedback.

Comment (cjerdonek) [reply Doc/library/asyncio-task.rst]: I don't know how soon the release will be cut, but you might want to consider merging now, and honing the docs in a follow-up PR. Otherwise, there's a chance you can miss the deadline.

Comment (aeros) [reply Doc/library/asyncio-task.rst]: > Adding .. warning:: section with a caution about blocking code inside a coroutine might be a good idea also.

That is certainly good asyncio advice, but I'm not certain that it would fit well with the rest of the current `asyncio.to_thread()` documentation. Perhaps we could consider placing something like that somewhere else, and linking to `asyncio.to_thread()`? 

To me, it seems like it would be more appropriately placed within a more general section that explains proper coroutine usage in asyncio.

**Review by methane — APPROVED** [review]

File: Doc/library/asyncio-task.rst
Comment (aeros) [reply Doc/library/asyncio-task.rst]: I'll go ahead and apply automerge now, good point @cjerdonek. I think the docs are now adequate after the last commit, we can improve further at a later time if needed.

**Comment by aeros** [comment]
Thanks for the review as well @methane. :-)

File: Doc/library/asyncio-task.rst
Comment (aeros) [reply Doc/library/asyncio-task.rst:608]: Done.

**Comment by aeros** [comment]
> Status check is done, and it's a failure x .

This occurred from a minor doctest warning, apparently having tildes in the code comments of example code is flagged as "suspicious". The latest commit should address it.

**Comment by zware** [comment]
Note that this missed the 3.9 boat; it's now merged into 3.10.  The whatsnew entry should be moved accordingly :)

**Comment by aeros** [comment]
> Note that this missed the 3.9 boat; it's now merged into 3.10. The whatsnew entry should be moved accordingly :)

Currently, I'm reaching out to Łukasz to see if we could consider including this in 3.9b2 since it was so close to the deadline (it's been a WIP since last year and would have otherwise made it on-time, if not for some last minute changes). If he decides that it's too late, I'll update the whatsnew entry accordingly.

File: Doc/library/asyncio-task.rst
Comment (asvetlov) [reply Doc/library/asyncio-task.rst]: Agree

**Comment by aeros** [comment]
Thanks, @ambv!

I'll try to open a PR to add the `versionadded` to the docs in the next couple of days, as @zware mentioned earlier.

File: Lib/asyncio/threads.py
Comment (JrooTJunior) [inline Lib/asyncio/threads.py:20]: I think this line should be changed to:
```suggestion
    context = contextvars.copy_context()
    func_call = functools.partial(context.run, functools.partial(func, *args, **kwargs))
```

because it can take unexpected behavior with contextvars.

For example i modify you'r example with adding usage of contextvars:
```python3
import asyncio
import time
from contextvars import ContextVar

ctx_value = ContextVar("ctx_value")


def blocking_io():
    print(f"start blocking_io at {time.strftime('%X')}")
    # Note that time.sleep() can be replaced with any blocking
    # IO-bound operation, such as file operations.
    time.sleep(1)
    print(f"value from context in blocking_io {ctx_value.get()}")
    print(f"blocking_io complete at {time.strftime('%X')}")


async def main():
    print(f"started main at {time.strftime('%X')}")

    ctx_value.set(42)

    await asyncio.gather(
        asyncio.to_thread(blocking_io),
        asyncio.sleep(1))

    print(f"finished main at {time.strftime('%X')}")


asyncio.run(main())
```

and got an error:
```
started main at 07:37:48
start blocking_io at 07:37:48
Traceback (most recent call last):
  File "/opt/project/asd.py", line 30, in <module>
    asyncio.run(main())
  File "/usr/local/lib/python3.10/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/opt/project/asd.py", line 25, in main
    await asyncio.gather(asyncio.to_thread(blocking_io), asyncio.sleep(1))
  File "/usr/local/lib/python3.10/asyncio/threads.py", line 21, in to_thread
    return await loop.run_in_executor(None, func_call)
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/project/asd.py", line 16, in blocking_io
    print(f"value from context in blocking_io {ctx_value.get()}")
LookupError: <ContextVar name='ctx_value' at 0x7f7ba2cb7130>
```
in this case i expects correct contextual value inside thread but currently it doesn't works as well

Comment (aeros) [reply Lib/asyncio/threads.py:20]: > in this case i expects correct contextual value inside thread but currently it doesn't works as well

I believe that is by design of `ContextVar`; in general, it is not shared across OS threads. `ContextVar` is primarily intended to be used within the same thread. For example, this also wouldn't work:

```py
from contextvars import ContextVar
import threading

ctx_value = ContextVar("ctx_value")

def worker():
    print(f"ctx_value in separate thread: {ctx_value.get()}")

def main():
    ctx_value.set(42)
    th = threading.Thread(target=worker)
    th.start()
    th.join()

if __name__ == '__main__':
    main()
```

I also think that in general that we don't want to implicitly share state across threads, as the above change would effectively do. However, we may want to consider adding a flag to `to_thread()` which allows for opt-in (`False` by default) passing of context state via something like the above. I personally don't know of the practical use cases for passing context to a separate thread though, so this will likely have to be a proposal in a separate bpo issue.

As the one who authored `contextvars` in the first place, what do you think of this @1st1?

Comment (1st1) [reply Lib/asyncio/threads.py:20]: In this case I think we should propagate the context. That's important for tracing APIs, decimal contexts etc. The only reason why this isn't already implemented for `run_in_executor()` is because it also allowed running process pools with which contexts don't play well.

The current implementation should be amended before 3.9beta2 to essentially do this:

```python
    loop = events.get_running_loop()
    context = contextvars.copy_context()
    func_call = functools.partial(context.run, func, *args, **kwargs)
    return await loop.run_in_context(None, func_call)
```

Please submit a PR, Kyle.

Comment (aeros) [reply Lib/asyncio/threads.py:20]: Okay, seems reasonable enough to me. I also need to add the `versionadded` to the docs anyways, I should have time later tonight to do that.

Edit: Note for myself later - add `Alex Root Junior <jroot.junior@gmail.com>` to the "Co-authored-by" for that PR for the contexvar suggestion.

Comment (aeros) [reply Lib/asyncio/threads.py:20]: This has been addressed in https://github.com/python/cpython/pull/20278, thanks for the feedback @JrooTJunior.

**Pull request merged**
